File: ml-platform/annotator/backend/app/api/tasks.py
"""Portal task APIs backed by service-authenticated platform endpoints."""

from datetime import datetime
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, Header, HTTPException, Query
from pydantic import BaseModel, ConfigDict, Field

from app.services.platform_client import PlatformClient, PlatformClientError
from app.services.session import PortalPrincipal, require_annotator_session

router = APIRouter(prefix="/portal", tags=["portal-tasks"])

class LabelWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")
    values: dict[str, Any]
    base_revision: int = Field(ge=0)

class BulkLabelWrite(BaseModel):
    model_config = ConfigDict(extra="forbid")
    items: list[dict[str, Any]] = Field(min_length=1)

class ConfirmationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    task_revision: int = Field(ge=0)
    scope_hash: str = Field(min_length=1, max_length=128)

def _failure(error: PlatformClientError) -> HTTPException:
    return HTTPException(status_code=error.status_code, detail=error.detail)

@router.get("/tasks")
async def list_tasks(
    cursor: str | None = None,
    limit: int = Query(default=50, ge=1, le=200),
    search: str | None = Query(default=None, max_length=128),
    status: str | None = Query(default=None, max_length=32),
    assignment_state: str | None = Query(default=None, max_length=32),
    sort: str = Query(default="created_at", pattern="^(created_at|due_at|status|assignment_state)$"),
    direction: str = Query(default="desc", pattern="^(asc|desc)$"),
    principal: PortalPrincipal = Depends(require_annotator_session),
):
    try:
        return await PlatformClient().internal_request(
            "GET",
            "/api/internal/portal/tasks",
            subject_id=str(principal.subject_id),
            scope="assignment:read",
            params={key: value for key, value in {
                "cursor": cursor,
                "limit": limit,
                "search": search,
                "status": status,
                "assignment_state": assignment_state,
                "sort": sort,
                "direction": direction,
            }.items() if value is not None},
        )
    except PlatformClientError as error:
        raise _failure(error) from error

@router.get("/tasks/{task_id}")
async def get_task(task_id: UUID, assignment_id: UUID | None = None, principal: PortalPrincipal = Depends(require_annotator_session)):
    try:
        return await PlatformClient().internal_request("GET", f"/api/internal/portal/tasks/{task_id}", subject_id=str(principal.subject_id), scope="assignment:read", params={"assignment_id": str(assignment_id)} if assignment_id else {})
    except PlatformClientError as error:
        raise _failure(error) from error

@router.get("/tasks/{task_id}/samples")
async def list_samples(task_id: UUID, assignment_id: UUID | None = None, cursor: str | None = None, limit: int = Query(default=50, ge=1, le=200), offset: int = Query(default=0, ge=0), sample_search: str | None = Query(default=None, min_length=1, max_length=256), label_status: str | None = Query(default=None, pattern="^(complete|incomplete)$"), comment_status: str | None = Query(default=None, pattern="^(open|resolved|none)$"), modified_after: datetime | None = Query(default=None), authorized_field: str | None = Query(default=None, min_length=1, max_length=128), authorized_value: str | None = Query(default=None, max_length=256), principal: PortalPrincipal = Depends(require_annotator_session)):
    try:
        return await PlatformClient().internal_request("GET", f"/api/internal/portal/tasks/{task_id}/samples", subject_id=str(principal.subject_id), scope="assignment:read", params={key: value for key, value in {"cursor": cursor, "limit": limit, "offset": offset if offset else None, "sample_search": sample_search, "label_status": label_status, "comment_status": comment_status, "modified_after": modified_after.isoformat() if modified_after else None, "authorized_field": authorized_field, "authorized_value": authorized_value, **({"assignment_id": str(assignment_id)} if assignment_id else {})}.items() if value is not None})
    except PlatformClientError as error:
        raise _failure(error) from error

@router.put("/tasks/{task_id}/samples/{sample_id}/labels")
async def save_labels(task_id: UUID, sample_id: str, data: LabelWrite, assignment_id: UUID | None = None, principal: PortalPrincipal = Depends(require_annotator_session)):
    try:
        return await PlatformClient().internal_request("PUT", f"/api/internal/portal/tasks/{task_id}/samples/{sample_id}/labels", subject_id=str(principal.subject_id), scope="assignment:write", json=data.model_dump(), params={"assignment_id": str(assignment_id)} if assignment_id else {})
    except PlatformClientError as error:
        raise _failure(error) from error

@router.post("/tasks/{task_id}/bulk-labels")
async def bulk_labels(task_id: UUID, data: BulkLabelWrite, assignment_id: UUID | None = None, principal: PortalPrincipal = Depends(require_annotator_session)):
    try:
        return await PlatformClient().internal_request("POST", f"/api/internal/portal/tasks/{task_id}/bulk-labels", subject_id=str(principal.subject_id), scope="assignment:write", json=data.model_dump(), params={"assignment_id": str(assignment_id)} if assignment_id else {})
    except PlatformClientError as error:
        raise _failure(error) from error

@router.post("/tasks/{task_id}/confirm")
async def confirm(task_id: UUID, data: ConfirmationRequest, assignment_id: UUID | None = None, principal: PortalPrincipal = Depends(require_annotator_session)):
    try:
        return await PlatformClient().internal_request("POST", f"/api/internal/portal/tasks/{task_id}/confirm", subject_id=str(principal.subject_id), scope="assignment:write", json=data.model_dump(), params={"assignment_id": str(assignment_id)} if assignment_id else {})
    except PlatformClientError as error:
        raise _failure(error) from error

@router.post("/tasks/{task_id}/edit-for-return")
async def edit_for_return(task_id: UUID, data: ConfirmationRequest, assignment_id: UUID | None = None, principal: PortalPrincipal = Depends(require_annotator_session)):
    try:
        return await PlatformClient().internal_request("POST", f"/api/internal/portal/tasks/{task_id}/edit-for-return", subject_id=str(principal.subject_id), scope="assignment:write", json=data.model_dump(), params={"assignment_id": str(assignment_id)} if assignment_id else {})
    except PlatformClientError as error:
        raise _failure(error) from error

@router.post("/tasks/{task_id}/return")
async def return_task(task_id: UUID, data: ConfirmationRequest, assignment_id: UUID | None = None, idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"), principal: PortalPrincipal = Depends(require_annotator_session)):
    if not idempotency_key:
        raise HTTPException(status_code=400, detail={"code": "IDEMPOTENCY_KEY_REQUIRED"})
    try:
        return await PlatformClient().internal_request("POST", f"/api/internal/portal/tasks/{task_id}/return", subject_id=str(principal.subject_id), scope="assignment:return", json=data.model_dump(), headers={"Idempotency-Key": idempotency_key}, params={"assignment_id": str(assignment_id)} if assignment_id else {})
    except PlatformClientError as error:
        raise _failure(error) from error

File: ml-platform/annotator/frontend/src/api/tasks.ts
import { request } from './client'
import { createUuid } from '../utils/uuid'

export type Task = {
  id: string
  assignment_id?: string
  title: string
  status: string
  state?: string
  due_at?: string
  completed_samples?: number
  total_samples?: number
  task_revision: number
  read_only?: boolean
  scope_hash: string
  samples?: Sample[]
  instructions?: string
  visible_columns?: string[]
  label_schema?: LabelSchema
}
export type LabelColumn = {
  machine_key: string
  display_name?: string
  value_type: 'int' | 'float' | 'string'
  required?: boolean
  enum_values?: Array<string | number>
  min_value?: number
  max_value?: number
  max_length?: number
}
export type LabelSchema = { columns: LabelColumn[] }
export type Sample = {
  sample_id: string
  values: Record<string, unknown>
  labels: Record<string, unknown>
  revision: number
}
export type SampleFilters = {
  sample_search?: string
  label_status?: 'complete' | 'incomplete'
  comment_status?: 'open' | 'resolved' | 'none'
  modified_after?: string
  authorized_field?: string
  authorized_value?: string
}

export type TaskQueueQuery = {
  cursor?: string
  limit?: number
  search?: string
  status?: string
  assignment_state?: string
  sort?: 'created_at' | 'due_at' | 'status' | 'assignment_state'
  direction?: 'asc' | 'desc'
}

export const listTasks = (query: TaskQueueQuery = {}) =>
  request<{ items: Task[]; total: number; next_cursor?: string | null }>('/portal/tasks', {
    query: { limit: 50, ...query },
  })
export const getTask = (id: string, assignmentId?: string) => request<Task>(`/portal/tasks/${id}`, { query: { assignment_id: assignmentId } })
export const listSamples = (id: string, cursor?: string, assignmentId?: string, filters: SampleFilters = {}, offset = 0) =>
  request<{ items: Sample[]; next_cursor?: string }>(`/portal/tasks/${id}/samples`, {
    query: { cursor, limit: 50, offset, assignment_id: assignmentId, ...filters },
  })
export const saveLabels = (
  taskId: string,
  sampleId: string,
  values: Record<string, unknown>,
  baseRevision: number,
  assignmentId?: string,
) =>
  request<{ values: Record<string, unknown>; revision: number; task_revision?: number }>(`/portal/tasks/${taskId}/samples/${sampleId}/labels`, {
    method: 'PUT',
    query: { assignment_id: assignmentId },
    body: JSON.stringify({ values, base_revision: baseRevision }),
  })
export const bulkLabels = (taskId: string, items: Array<Record<string, unknown>>, assignmentId?: string) =>
  request<{ items: Array<{ sample_id: string; values: Record<string, unknown>; revision: number }> }>(`/portal/tasks/${taskId}/bulk-labels`, {
    method: 'POST',
    body: JSON.stringify({ items }),
    query: { assignment_id: assignmentId },
  })
export const confirmTask = (taskId: string, taskRevision: number, scopeHash: string, assignmentId?: string) =>
  request(`/portal/tasks/${taskId}/confirm`, {
    method: 'POST',
    body: JSON.stringify({ task_revision: taskRevision, scope_hash: scopeHash }),
    query: { assignment_id: assignmentId },
  })
export const editForReturn = (taskId: string, taskRevision: number, scopeHash: string, assignmentId?: string) =>
  request(`/portal/tasks/${taskId}/edit-for-return`, {
    method: 'POST',
    body: JSON.stringify({ task_revision: taskRevision, scope_hash: scopeHash }),
    query: { assignment_id: assignmentId },
  })
export const returnTask = (taskId: string, taskRevision: number, scopeHash: string, assignmentId?: string) =>
  request(`/portal/tasks/${taskId}/return`, {
    method: 'POST',
    headers: { 'Idempotency-Key': createUuid() },
    body: JSON.stringify({ task_revision: taskRevision, scope_hash: scopeHash }),
    query: { assignment_id: assignmentId },
  })
